### PR TITLE
ICU-21284 More MeasureFormatTest and NumberFormatterApiTest test cases

### PR DESCRIPTION
**Pull request by hugovdm**
##### Checklist

- [X] Issue filed: https://unicode-org.atlassian.net/browse/ICU-21284
- [X] Updated PR title and link in previous line to include Issue number
- [X] Issue accepted
- [X] Tests included
- N/A Documentation is changed or added
- [x] Decide what to do with MeasureFormat (leave as is, or update docs a bit? Test what it can and cannot handle?)
- [x] Test inverse input unit behaviour - **_taken care of by #1565_**
- [x] Test whether (almost) all unit identifiers in the validity file can be parsed

Closes: https://github.com/icu-units/icu/issues/38

### CODE REVIEW

File: icu4c/source/i18n/unicode/measfmt.h [hunk 1]
```diff
@@ -91,7 +91,8 @@ class DateFormat;
 /**
  * <p><strong>IMPORTANT:</strong> New users are strongly encouraged to see if
  * numberformatter.h fits their use case.  Although not deprecated, this header
- * is provided for backwards compatibility only.
+ * is provided for backwards compatibility only, and has much more limited
+ * capabilities.
```
Comment (hugovdm) [inline icu4c/source/i18n/unicode/measfmt.h:95]: A minor touching of unicode/*.h ...

Maybe it should be a different PR? It's here mostly because it was in response to considering unit tests for MeasureFormat, but then recognising its use is strongly discouraged.

File: icu4j/main/classes/core/src/com/ibm/icu/util/MeasureUnit.java [hunk 2]
```diff
@@ -695,7 +695,8 @@ public boolean equals(Object rhs) {
      */
     @Override
     public String toString() {
-        return type + "-" + subType;
+        String result = measureUnitImpl == null ? type + "-" + subType : measureUnitImpl.getIdentifier();
+        return result == null ? "" : result;
```
Comment (hugovdm) [inline icu4j/main/classes/core/src/com/ibm/icu/util/MeasureUnit.java:699]: Bringing in getIdentifier() logic. Else this returns "null-null", noticed in debug / unit test output.

Comment (hugovdm) [reply icu4j/main/classes/core/src/com/ibm/icu/util/MeasureUnit.java:699]: ("Type" might be invalid for not-built-in-types, so I'm keeping "type" for built-ins and dropping it otherewise...)

File: icu4j/main/tests/core/src/com/ibm/icu/dev/test/format/MeasureUnitTest.java [hunk 3]
```diff
@@ -3596,7 +3604,6 @@ public void TestInvalidIdentifiers() {
         }
     }
 
-    @Test
```
Comment (hugovdm) [inline icu4j/main/tests/core/src/com/ibm/icu/dev/test/format/MeasureUnitTest.java:3599]: Whoops. Fixing.

**Review by sffc — APPROVED** [review]

**Comment by sffc** [comment]
LGTM except check the failing CI tests.

**Comment by hugovdm** [comment]
Looks like we have a NaN beahvioural difference on MSVC:

            unitUsage {
            FAIL: NaN with Unit Preferences: Unsafe Path; got "-NaN cm\u00B2"; expected "NaN cm\u00B2"
            FAIL: NaN with Unit Preferences: Safe Path; got "-NaN cm\u00B2"; expected "NaN cm\u00B2"
            FAIL: NaN with Unit Preferences: Skeleton Path: 'unit/acre usage/default': nan; got "-NaN cm\u00B2"; expected "NaN cm\u00B2"
            FAIL: NaN with Unit Preferences: Concise Skeleton Path: 'unit/acre usage/default': nan; got "-NaN cm\u00B2"; expected "NaN cm\u00B2"
            FAIL: Negative numbers: NaN minute-and-second: Unsafe Path; got "-NaN sec"; expected "NaN sec"
            FAIL: Negative numbers: NaN minute-and-second: Safe Path; got "-NaN sec"; expected "NaN sec"
            FAIL: Negative numbers: NaN minute-and-second: Skeleton Path: 'unit/second usage/media': nan; got "-NaN sec"; expected "NaN sec"
            FAIL: Negative numbers: NaN minute-and-second: Concise Skeleton Path: 'unit/second usage/media': nan; got "-NaN sec"; expected "NaN sec"

**Comment by sffc** [comment]
Yeah, MSVC retains the sign bit in NaN whereas other platforms typically do not.  A way around this is to use a NaN constant such as the one returned by `uprv_getNaN()`.

**Comment by hugovdm** [comment]
I'm already using uprv_getNaN() in the tests.

I also looked at the ComplexUnitsConverter::convert code to try to think where the NaN behaviour might differ. (And for some reason, the tests still succeeded on some platforms with MSVC.)

For now, I've commented out these tests, with a TODO to revisit. Rationale: the results aren't as misleading as returning "0", so "edge case bug", not ICU69 feature-related.

**Review by sffc — DISMISSED** [review]
MSVC is happy now.  LGTM please squash.

**Comment by hugovdm** [comment]
@sffc @younies this one's been approved by sffc, just needs a re-approve post squash.

**Review by sffc — APPROVED** [review]
